media-proxy/src/session-rdma-tx.cc

```diff
@@ -0,0 +1,137 @@
+    if (err != MEMIF_ERR_SUCCESS)
+        INFO("memif_refill_queue: %s", memif_strerror(err));
+}
+
+void TxRdmaSession::frame_thread()
+{
+    while (!atomic_load_explicit(&shm_ready, std::memory_order_acquire) && !stop)
+        usleep(1000);
+
+    INFO("%s, TX RDMA thread started\n", __func__);
+    while (!stop) {
+        if (!atomic_load_explicit(&shm_ready, std::memory_order_acquire))
+            continue;
+        handle_sent_buffers();
+    }
+}
+
+TxRdmaSession::TxRdmaSession(libfabric_ctx *dev_handle, const mcm_conn_param &request,
+                             memif_ops_t &memif_ops)
+    : Session(memif_ops, request.payload_type, TX), ep_cfg{0}, ep_ctx(0), stop(false),
+      frame_thread_handle(0), fb_send(0)
+{
+    transfer_size = request.payload_args.rdma_args.transfer_size;
+
+    ep_cfg.rdma_ctx = dev_handle;
+    memcpy(&ep_cfg.remote_addr, &request.remote_addr, sizeof(request.remote_addr));
+    memcpy(&ep_cfg.local_addr, &request.local_addr, sizeof(request.local_addr));
+    ep_cfg.dir = direction::TX;
+}
+
+int TxRdmaSession::init()
+{
+    int err = ep_init(&ep_ctx, &ep_cfg);
+    if (err) {
+        ERROR("Failed to initialize libfabric's end point");
+        return -1;
+    }
+
+    err = shm_init(transfer_size, 4);
+    if (err < 0) {
+        ERROR("Failed to to initialize shared memory");
+        return -1;
+    }
+
+    frame_thread_handle = new std::thread(&TxRdmaSession::frame_thread, this);
+    if (!frame_thread_handle) {
+        ERROR("Failed to create thread");
+        return -1;
+    }
+    return 0;
+}
+
+TxRdmaSession::~TxRdmaSession()
+{
+    INFO("%s, fb_send %d\n", __func__, fb_send);
+    stop = true;
+    if (ep_ctx) {
+        ep_destroy(&ep_ctx);
+        ep_ctx = 0;
+    }
+
+    if (frame_thread_handle) {
+        frame_thread_handle->join();
+        delete frame_thread_handle;
+        frame_thread_handle = 0;
+    }
+}
+
+int TxRdmaSession::on_receive_cb(memif_conn_handle_t conn, uint16_t qid)
+{
+    memif_buffer_t shm_bufs = {0};
+    uint16_t buf_num = 0;
+    int err = 0;
+
+    if (stop) {
+        INFO("TX session already stopped.");
+        return -EINVAL;
+    }
+
+    /* receive packets from the shared memory */
+    err = memif_rx_burst(conn, qid, &shm_bufs, 1, &buf_num);
+    if (err != MEMIF_ERR_SUCCESS && err != MEMIF_ERR_NOBUF) {
+        INFO("memif_rx_burst: %s", memif_strerror(err));
+        return err;
+    }
+
+    err = ep_send_buf(ep_ctx, shm_bufs.data, shm_bufs.len);
+    if (err) {
+        ERROR("ep_send_buf failed with: %s", fi_strerror(-err));
+        return err;
+    }
+
+    return 0;
+}
+
+int TxRdmaSession::on_connect_cb(memif_conn_handle_t conn)
+{
+    memif_region_details_t region;
+
+    int err = memif_get_buffs_region(conn, &region);
+    if (err) {
+        ERROR("%s, Getting memory buffers from memif failed. \n", __func__);
+        return err;
+    }
+
+    err = ep_reg_mr(ep_ctx, region.addr, region.size);
+    if (errno) {
```

Comment:
```suggestion
    if (err) {
```

Reply:
fix in https://github.com/OpenVisualCloud/Media-Communications-Mesh/pull/240/commits/2969a18f4536ff15a4531a07715199489671437e